odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1471 @@
+                                grad_lipschitz=0)
+        elif scalar < 0:
+            Functional.__init__(self, domain=func.domain,
+                                linear=func.is_linear,
+                                smooth=func.is_smooth,
+                                concave=func.is_convex,
+                                convex=func.is_concave,
+                                grad_lipschitz=(-scalar * func.grad_lipschitz))
+        else:
+            # It should not be possible to get here
+            raise TypeError('comparison with scalar {} failed'.format(scalar))
+
+        OperatorLeftScalarMult.__init__(self, operator=func, scalar=scalar)
+
+        self._func = func
+        self._scalar = scalar
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluated
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+
+        functional = self
+
+        class LeftScalarMultGradient(Operator):
+            """The gradient operator for a `FunctionalLeftScalarMult`
+            functional.
+            """
+            def __init__(self):
+                """Initialize a `LeftScalarMultGradient` instance."""
+                super().__init__(functional.domain, functional.domain,
+                                 linear=False)
+
+            def _call(self, x):
+                """Evaluates the gradient in a point ``x``.
+
+                Parameters
+                ----------
+                x : `LinearSpaceVector`
+                    Element in the domain of the operator.
+
+                Returns
+                -------
+                out : `LinearSpaceVector`
+                    The gradient in the point ``x``. An element in the doimain
+                    of the operator.
+                """
+                return functional._scalar * functional._func.gradient(x)
+
+        return LeftScalarMultGradient()
+
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of the scaled functional."""
+        # The helper function only allows positive scaling parameters.
+        # Otherwise it gives an error.
+        return ConvexConjugateFuncScaling(
+            self._func.conjugate_functional, self._scalar)
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the scaled functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator.
+
+        Returns
+        -------
+        out : `Operator`
+            Domain and range equal to domain of functional.
+        """
+        sigma = float(sigma)
+        if sigma * self._scalar < 0:
```

Comment:
Technically one can create a convex functional, scale it with a negative value, and then compute the proximal operator with a negative step length since scalar*step length will be positive. Is this something we want to keep, or should scaling with negative scalars automatically disqualify the use of proximal operators?